lib/Arch.cpp

```diff
@@ -0,0 +1,441 @@
+      // figure out unsigned vs signed ints
+      auto sign = true;
+      switch (derived.getBitWidth()) {
+        case 8: {
+          ret = sign ? "b" : "B";
+          break;
+        }
+        case 16: {
+          ret = sign ? "h" : "H";
+          break;
+        }
+        case 32: {
+          ret = sign ? "i" : "I";
+          break;
+        }
+        case 64: {
+          ret = sign ? "l" : "L";
+          break;
+        }
+      }
+      break;
+    }
+    case llvm::Type::FunctionTyID: {
+      ret = "func";
+      break;
+    }
+    case llvm::Type::StructTyID: {
+      auto struct_ptr = llvm::cast<llvm::StructType>(&type);
+      std::string element_list = "";
+      for (unsigned i = 0; i < struct_ptr->getNumElements(); i++) {
+        // TODO: BUG: this if statement protects against infinite chained
+        // pointers, which this code has a hard time with. Notably, this
+        // includes some types of C++ ostream functions.
+        if (struct_ptr->getElementType(i)->isPtrOrPtrVectorTy()) {
+          element_list += "?";
+          continue;
+        }
+        element_list += TranslateType(*struct_ptr->getElementType(i));
+      }
+      ret += " {" + element_list + "}";
+
+      break;
+    }
+    case llvm::Type::ArrayTyID: {
+      ret = "array";
+      break;
+    }
+    case llvm::Type::PointerTyID: {
+      ret = "*";
+      auto derived = llvm::dyn_cast<llvm::PointerType>(&type);
+      // Get the type of the pointee
+      ret += TranslateType(*derived->getElementType());
+      break;
+    }
+
+    default:
+      LOG(ERROR) << "Could not translate TypeID: " << id;
+      break;
+  }
+  return ret;
+}
+
+// Try to allocate a register for the argument based on the register constraints
+// and what has already been reserved. Return nullptr if there is no possible
+// register allocation.
+remill::Register *TryRegisterAllocate(
+    llvm::Type &type, std::vector<bool> &reserved,
+    const std::vector<RegisterConstraint> &register_constraints) {
+  SizeConstraint size_constraint;
+  TypeConstraint type_constraint;
+
+  if (type.isIntegerTy()) {
+    type_constraint = kTypeInt;
+    auto derived = llvm::cast<llvm::IntegerType>(type);
+    unsigned int width = derived.getBitWidth();
+    if (width == 64) {
+      size_constraint = kMinBit64;
+    } else {
+      // TODO: I know that this is wrong but for now its fine
+      size_constraint = kMinBit32;
+    }
+  } else if (type.isFloatTy()) {
+    type_constraint = kTypeFloat;
+    // We automatically know it is 32-bit IEEE floating point type
+    size_constraint = kMinBit32;
+  } else if (type.isDoubleTy()) {
+    type_constraint = kTypeFloat;
+    // We automatically know it is 64-bit IEEE floating point type
+    size_constraint = kMinBit64;
+  } else if (type.isPointerTy()) {
+    type_constraint = kTypeIntegral;
+    size_constraint = kMinBit64;
+  }
+  // TODO: Handle other types
```

Comment:
`TODO(aaronyoo):`.

You should handle `long double`s, i.e. the `X86_FP80` type here. Possibly also the MMX type.

Reply:
If you leave something undone, add a `LOG(ERROR)` or in this case, maybe even a `LOG(FATAL)`. Right now this is silently ignoring, which is not good because we won't know when we're missing stuff.

Reply:
Left a TODO for myself to handle MMX types 327d728